fusion.md

```diff
@@ -0,0 +1,1396 @@
+object is not a container (i.e., has the `SIMPLE` flag, then the match will be
+rejected even if the data name matches.  
+
+In addition if no data name and no filter is set then not only does this
+FilterStep match all data but it also enables special "descendant step"
+traversal behavior in GCAMFusion where the next FilterStep can be matched zero
+or more containers down.  This is analogous to the `//` operator in XPath
+queries.  For example, if a 'sector' object has 'subsector' children, which in
+turn have 'technology' children of their own, then `sector//share-weight` will
+find data named 'share-weight' at both the subector and technology levels.  It
+would also find a share-weight object contained in the sector itself if there
+were any, but in this example there are no such matches; share weights are only
+defined for subsectors and technologies.
+<!-- XXX: Check if this last sentence is accurate: would sector//share-weight -->
+<!-- match a share weight that is a direct child of the sector (if there were -->
+<!-- one)? -->
+
+
+
+### Filter objects
+
+A Filter object allows GCAMFusion to select a subset of a data object that is a
+container for other objects. The available Filters are:
+
+* NoFilter: matches all elements
+* NamedFilter: Calls the -&gt;getName() method on the container to compare in it's predicate. `SIMPLE` and `ARRAY` will never match.
+* YearFilter: Calls the -&gt;getYear() method if the data is `CONTAINER` or uses the `Modeltime` to convert period index to year if the data is `ARRAY` and compares that year in it's predicate. `SIMPLE` data will never match.
+* IndexFilter: Computes the index offset of each element in an `ARRAY` or `CONTAINER` Data for comparison in it's predicate. `SIMPLE` Data will never match.
+
+
+
+### Predicates
+
+
+
+A predicate is a way to test whether a year, or name, or index, etc matches a
+value the user was looking for. Currently predicates can only operate on string
+and int. If a predicate that is doing a string comparison is given an int to
+match (i.e. called from a YearFilter) it will always return false, and vice
+versa. The available predicates are:
+<!-- XXX: Is it true that an int predicate will always fail to match a string? -->
+
+|   StringEquals | string | Tests if the proposition exactly matches a string value.|
+|   StringRegexMatches | string | Tests if the proposition matches a regular expression in the egrep notation.|
+|   IntEquals | int | Tests if the proposition exactly matches an int value.|
+|   IntGreaterThan | int | Tests if the proposition is strictly greater than an int value.|
+|   IntGreaterThanEq | int | Tests if the proposition is greater or equal to an int value.|
+|   IntLessThan | int | Tests if the proposition is strictly less than an int value.|
+|   IntLessThanEq | int | Tests if the proposition is less or equal to an int value.|
+
+<!-- XXX: which notation do we want to use, check C++ standard for options -->
```

Comment:
This question was originally attached to the `StringRegexMatches`.  I think we should use whatever the default is for the regular expression engine we are using.

Reply:
I have it configured as the "egrep" syntax.  I'll update the documentation to clarify (unless you want to use some other type of syntax):
http://www.cplusplus.com/reference/regex/regex_constants/